proposals/type-parameter-inference-from-constraints.md

```diff
@@ -0,0 +1,141 @@
+now is an excellent time to re-examine the concern on the breaking change here, and potentially move forward with the proposal.
+
+Credit to [@HellBrick](https://github.com/HellBrick) for the [original proposed mechanics](https://github.com/dotnet/roslyn/issues/5023#issuecomment-154728796) of the design.
+This proposal has been further refined from their original starting point.
+
+### Use cases this supports:
+
+```csharp
+private static void M<T, X>(T Object) where T : IEnumerable<X>, IComparable<X> 
+{
+}
+
+private class MyClass : IComparable<String>, IEnumerable<String> 
+{
+}
+
+private static void CallMyFunction() 
+{
+    var c = new MyClass();
+    M(c);
+}
+```
+
+## Detailed design
+
+### Changes to Type Inference Algorithm
+
+We modify the type inference process described in [§12.6.3](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#1263-type-inference) to include
+constraint relationships in the dependence relationship between type variables.
+
+#### Enhanced Dependence Relationship - Modified Spec Text
+
+The following text from [§12.6.3.6 Dependence](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#12636-dependence) is modified:
+
+> **12.6.3.6 Dependence**
+>
+> An *unfixed* type variable `Xᵢ` *depends directly on* an *unfixed* type variable `Xₑ` if one of the following holds:
+>
+> - For some argument `Eᵥ` with type `Tᵥ` `Xₑ` occurs in an *input type* of `Eᵥ` with type `Tᵥ` and `Xᵢ` occurs in an *output type* of `Eᵥ` with type `Tᵥ`.
+> - **`Xᵢ` occurs in a constraint for `Xₑ`.**
+>
+> `Xₑ` *depends on* `Xᵢ` if `Xₑ` *depends directly on* `Xᵢ` or if `Xᵢ` *depends directly on* `Xᵥ` and `Xᵥ` *depends on* `Xₑ`. Thus "*depends on*" is the transitive but not reflexive closure of "*depends directly on*".
+
+#### Enhanced Fixing Process - Modified Spec Text
+
+The following text from [§12.6.3.12 Fixing](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#126312-fixing) is modified:
+
+> **12.6.3.12 Fixing**
+>
+> An *unfixed* type variable `Xᵢ` with a set of bounds is *fixed* as follows:
+>
+> - The set of *candidate types* `Uₑ` starts out as the set of all types in the set of bounds for `Xᵢ`.
+> - Each bound for `Xᵢ` is examined in turn: For each exact bound U of `Xᵢ` all types `Uₑ` that are not identical to `U` are removed from the candidate set. For each lower bound `U` of `Xᵢ` all types `Uₑ` to which there is *not* an implicit conversion from `U` are removed from the candidate set. For each upper-bound U of `Xᵢ` all types `Uₑ` from which there is *not* an implicit conversion to `U` are removed from the candidate set.
+> - If among the remaining candidate types `Uₑ` there is a unique type `V` to which there is an implicit conversion from all the other candidate types, then `Xᵢ` is fixed to `V` **and a lower-bound inference is performed from `V` to each of the types in `Xᵢ`'s constraints, if any**.
+> - Otherwise, type inference fails.
```

Comment:
I haven't thought about the full implications of this algorithm, namely the set of things that will now be inferrable, and what will remain uninferrable.

Ideally I think we'd do as much as possible here. By that I mean, a general statement that if it is theoretically possible to infer the type, we do so. In practice, I'm worried our subtyping + overload resolution rules will make this undecidable.

But it's worth digging in to the literature to figure out precisely what will fall over.

Reply:
Btw, starting w/ Algorithm W is not a bad idea.

Reply:
I've thought about this for a bit and my new thoughts are: I don't think this change is very complicated, and I'm not sure we have much to worry about.

Basically, we're not doing full type inference, we're just doing type subsitution for particular invocations. It doesn't seem like there's a lot of information that we could be missing with this algorithm.